### PR TITLE
Generalize host allocator to be device-agnostic

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123080
* __->__ #123079

# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImpl` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `CachingHostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImpl` as its implementation.
- `CUDACachingHostAllocator` is a CUDA host allocator whose implementation is `CUDACachingHostAllocatorImpl` which is specialized from `CachingHostAllocatorImpl`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/core/CachingHostAllocator.h [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+ * consumed in the free path.
+ *   4) In the empty_cache path, we flush any available blocks into the free
+ * list. Remove all element of free list, then remove them from block list and
+ * release the associated pinned memory allocation via free_block.
+ *
+ * To share the mechanism across each backend, we abstract its logic to be
+ * device-agnostic through an interface. Then each backend can specialize
+ * their own implement and the related runtime functions, like
+ * allocate_host_memory, free_block, record_stream, query_event; Even inheriting
+ * from the interface to extend additional features.
```
Comment (jgong5) [inline aten/src/ATen/core/CachingHostAllocator.h:79]: Can we make it more clear what a backend implementation needs to do, e.g., creating a class `xyz` that derives `CachingHostAllocatorImplInterface`, specifying template args (can you add explanation to the temp args too?), implementing methods `abc`... ,etc.?

**Review by jgong5 — APPROVED** [review]
Overall LGTM. Suggest to add more notes on what a backend needs to do to add its own host allocator.

File: aten/src/ATen/core/CachingHostAllocator.h (hunk 1 above)
Comment (guangyey) [reply aten/src/ATen/core/CachingHostAllocator.h]: refine comments and add more notes.

File: aten/src/ATen/core/CachingHostAllocator.h [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+ * Note that this caching host allocator does not split larger allocations into
+ * smaller blocks, unlike the caching device allocator.
+ */
+
+template <
+    typename S,
+    typename E,
+    typename B = HostBlock<S>,
+    typename C = ComparatorSize<B>>
+struct CachingHostAllocatorImplInterface {
```
Comment (EikanWang) [inline aten/src/ATen/core/CachingHostAllocator.h:116]: `Caching` is the detailed implementation. Should we just rename the class name as `HostAllocatorImplInterface`?

**Review by EikanWang — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Is this actually used by default on CUDA today? Do you plan on using it by default for XPU?

Also reviewed assuming all the code was just moved. Did you actually do any logic change beyond adding the appropriate API? If not, SGTM!

**Comment by guangyey** [comment]
> Is this actually used by default on CUDA today? Do you plan on using it by default for XPU?
> 
> Also reviewed assuming all the code was just moved. Did you actually do any logic change beyond adding the appropriate API? If not, SGTM!

Yes, this caching mechanism is used on CUDA by default, and XPU will follow it.
I do **NOT** change any logic, just add some appropriate APIs to decouple its implementation with CUDA. To make another backend, like XPU, can share these pieces of code.

**Review by gujinghui — APPROVED** [review]

**Pull request closed without merging**